### PR TITLE
Template jsonfilereader

### LINKED ISSUE #28850: Handle string data in JSONFileReader
### Motivation

I'm trying to build objects in an Action based on data in a JSON file. Some of these objects take string parameters/options or have block restrictions by element set name. The JSONFileReader UO only handles Real data.

### Design

Add extra functions to JSONFileReader to handle string data. Current functionality needs to be able to handle returning both single value and vector of values, as well as the input being either a single key or a vector or keys, so 4 functions total. Having string-based options would double that. May be worth looking into templating the functions. One may also want to handle integers too.

### Impact

Something new. May change existing API's if templating things.

### PR DESCRIPTION
**Pull request by jessecarterMOOSE**
## Reason
Only Real data type currently supported when reading from JSON file with `JSONFileReader` UO.

## Design
Changed getter functions to use templates.

## Impact
Will change existing API's in that the data type will need to be specified. Not sure how many downstream apps this will break. The original getters could probably be restored if necessary.

Note I don't have much experience with templated functions in C++ and had to implement the function source in the header files to get things to link properly. If there is a better way then let me know. Also I updated unit tests but don't know how to build and run them.

closes #28850

### CODE REVIEW

**Comment by jessecarterMOOSE** [comment]
tagging @GiudGiud based on our Slack discussion the other day

File: framework/include/userobjects/JSONFileReader.h [hunk 1]
```diff
@@ -36,30 +36,94 @@ class JSONFileReader : public GeneralUserObject
    * @param scalar_name the name of the desired scalar
    * @param scalar reference to the scalar that will be set with the scalar in the JSON
    */
-  void getScalar(const std::string & scalar_name, Real & scalar) const;
+  template <typename T>
+  void getScalar(const std::string & scalar_name, T & scalar) const
```
Comment (GiudGiud) [inline framework/include/userobjects/JSONFileReader.h:40]: can we "getScalar" for a string or does that sound weird.

if so, should it be getSingleton ?

Comment (jessecarterMOOSE) [reply framework/include/userobjects/JSONFileReader.h:40]: I think that sounds even weirder 🤔

What about getValue and getVectorValue?

Comment (GiudGiud) [reply framework/include/userobjects/JSONFileReader.h:40]: haha
yes that sounds good

**Review by GiudGiud — APPROVED** [review]
for extra credit I think we should have a custom error when the data exists in the JSON file but is of a different type

**Comment by GiudGiud** [comment]
Test failures are unrelated

**Comment by jessecarterMOOSE** [comment]
thanks @GiudGiud

**Pull request merged**